### ISSUE TITLE
Enable external CalDAV calendar read-write.

### ISSUE THREAD

**Issue by simpledissection** [nextcloud/calendar#2766]
**Is your feature request related to a problem? Please describe.**
It is frustrating that external calendars can only be added for viewing.  For my use case, that functionality limitation impedes my family's entire ability to use the Nextcloud Calendar app

**Describe the solution you'd like**
I would like to see an option implemented to add external calendars via a link that will enable calendar synchronization with an external CalDAV server.  This would extend the functionality of the NextCloud app as a true one-stop-shop calendar where a user could both create and have full access to any calendars they need, including existing calendars from other services.

**Describe alternatives you've considered**
Honestly, I can't think of any alternatives that would provide the needed functionality.  I have already attempted using various third party applications and displaying them as 'external sites', however, so far, none provide the type of functionality that such a feature would enable if added to the Nextcloud Calendar.  Aside from that, using the external sites requires an additional login, where added functionality could maintain saved credentials.

**Additional context**
Here is an idea of what it might look like in the interface:
![image](https://user-images.githubusercontent.com/8690712/103178390-70ae8080-4850-11eb-9bc9-c880ff009ebb.png)

**Comment by dagguh** [comment]
This is an important step when gradually de-Googling your personal life, family, friends and work.

**Comment by silopolis** [comment]
That'd surely be a major enhancement !

**Comment by miguelarios** [comment]
Yes! I want Nextcloud to be my calendar client just as it can be my mail client.

**Comment by ant0nwax** [comment]
I vote +1 if I may vote here, otherwise I would not be able to use nextcloud easily
a calendar nowadays runs on the same system most times like the mail system in my honest opinion
i will anyway make tries with using the internal nextcloud calendar instead of the one which is paired with the email inbox

**Comment by rasos** [comment]
SabreDAV also has a php [library for a WebDAV/CalDAV/DardDAV client](https://sabre.io/dav/davclient/). This may be a starting point for nextcloud acting not only as a CalDAV server, but also as a client, even if the [code on github](https://github.com/sabre-io/davclient) seems quite outdated.

**Comment by rasos** [comment]
We might try to re-use parts of the Calcom's CalDav server code https://github.com/calcom/cal.com  

Cal.com is a service for appointments handling and can play the client role for several calendar services. 

![calcomServers](https://user-images.githubusercontent.com/485723/199953779-6db4280f-57bc-4cd8-af76-d6bdc60c4d33.png)

Cal.com code is GPLv3 and uses:

- [Next.js](https://nextjs.org/) with TypeScript
- [React](https://reactjs.org/)
- [Tailwind](https://tailwindcss.com/)
- [Prisma](https://prisma.io/)

Code to handle CalDAV calendar is in calcom/packages/app-store/caldavcalendar and in calcom/packages/lib/CalendarService.ts

**Comment by tcitworld** [comment]
That's not especially helpful, but thanks anyway.

The Sabre DAV client is already used at Nextcloud for pretty much the same things with WebDAV (basis of the federation feature) and the address book sync. It will have to get extended before anyway to support the `REPORT` WebDAV method.

As this issue is a required basis for Calendar sharing federation https://github.com/nextcloud/server/issues/1440, I'm starting to look at it more closely, but there's really a lot to do. :)

**Comment by rasos** [comment]
The [tsdav library](https://github.com/natelindev/tsdav)  might be another candidate to fetch calendar events via CalDav into nextcloud. Has been tested with various cloud providers.

**Comment by battosai30** [comment]
+1 for this feature.

**Comment by moovinghead** [comment]
+1 for this feature from me as well

**Comment by mschilli87** [comment]
My current workaround is to run vdirsyncer via chron on the same server I am hosting my Nextcloud instance on to syncronize the remote calendar with a local copy of that calendar. Then, on all my devices I just sync all my Nextcloud calendars pushing and pulling changes to the remote calendar via its local copy. It takes a bit more work to setup properly and depending on the remote calendar might require an addional account for write access but at least I get read/write access to shared calendars via Nextcloud and khal. Just figured I share my approach in case it was useful for anyone else while waiting for this feature.

**Comment by ant0nwax** [comment]
Thanks for sharing, would it be a big thing if I request a draw.io diagram of your calendar setup? From the lines until now I hardly can imagine how many calendars are involved here for one account.

**Comment by mschilli87** [comment]
@ant0nwax
> Thanks for sharing, would it be a big thing if I request a draw.io diagram of your calendar setup?

This is the best I could come up with:

![calendars](https://user-images.githubusercontent.com/12913701/226214005-b2b3bd3e-943e-4858-a689-09ebb01c5504.svg)

I hope it helps.

**Comment by ant0nwax** [comment]
@mschilli87 I am speechless.... without words
THANK VERY VERY MUCH, this is awesome!
Is the vdirsyncer on github and is it this repo? https://github.com/pimutils/vdirsyncer

**Comment by mschilli87** [comment]
@ant0nwax
> Is the vdirsyncer on github and is it this repo? https://github.com/pimutils/vdirsyncer

Yes, that's the one. They provide sufficient documentation on how to configure sync jobs with various servers.

**Comment by nanaksingh13** [comment]
+1
Any updates on this feature.

**Comment by dohlin** [comment]
> @ant0nwax
> 
> > Thanks for sharing, would it be a big thing if I request a draw.io diagram of your calendar setup?
> 
> This is the best I could come up with:
> 
> ![calendars](https://user-images.githubusercontent.com/12913701/226214005-b2b3bd3e-943e-4858-a689-09ebb01c5504.svg)
> 
> I hope it helps.

Just wanted to say thank you for this.  Took me a little bit to get figured out, and I definitely don't feel like it should be this hard...but this workaround seems to work really, really well.  So thanks again!!

**Comment by gitwittidbit** [comment]
+1 for the read/write client feature

**Comment by FFricke** [comment]
a big big big +1 from me

**Comment by xtrails101** [comment]
For me it will be a good solution, if I can add a Calendar (&Contacts) from my OpenExchange to my nextcloud with read/write access. So I have Mail, Calendar & Contacts on OpenExchange but can use it through one interface in combination with my files... on my nextcloud.
Actually only read is supported, maybe my +1 will push this enhancement a little bit ;-)

**Comment by alexanderdd** [comment]
can someone triage this? (and remove the label to triage?)

Do I understand correctly that first this needs to be implemented https://github.com/nextcloud/server/issues/1440 and then the feature has to be implemented in calendar, which this issue is for?

**Comment by tcitworld** [comment]
It's the other way around. This issue is part of having https://github.com/nextcloud/server/issues/1440.

**Comment by ThHirsch** [comment]
+1 I would also like to have a full caldav sync with read/write. So NC can be my primary Interface to handle my email and calendar.

**Comment by alexanderdd** [comment]
@ThHirsch please react with 👍 to the first post instead of a comment with +1, this is our way in all Nextcloud repositories (because then we can sort by number of 👍 )

**Comment by ThHirsch** [comment]
@alexanderdd Thank you for that info. Just added my  👍 to the first post and wil step back from +1 posts in the future.

**Comment by Spacelord09** [comment]
+1 Any updates to this?

**Comment by mschilli87** [comment]
@Spacelord09: If so, you'd find it here. If you can't help, please don't slow down developers by making them read your comments that add nothing.

**Comment by luposlip** [comment]
I'm in the process of getting out of all bigtech (incl. github, BTW). And the calendar app in Nextcloud is nice. But I can't subscribe to my CalDAV calendar, my CalDAV principal calendars or WebDAV calendar, which require username and password.

This is a shame, because it means I still have to check several calendars, to figure out if I'm available and not. Furthermore I can't reliably use the Free/Busy feature of the NextCloud calendar, because it only includes half my calendars.

**Comment by bremme** [comment]
+1 for me as well on this issue. Currently in the process of migrating my contacts and calendar from Google to Nextcloud. This seems like a deal breaker. I need to have write access to different shared calendars. Not sure, maybe a I find a way around it if I do more research. But my first Google search landed me here.

**Comment by mschilli87** [comment]
@bremme: Excuse the plug but as it is hidden as 'old' by GitHub by now, have a look at my workaround that I use for years: https://github.com/nextcloud/calendar/issues/2766#issuecomment-1475421159. It is ugly but has allowed me to collaborate with others while having all my data on my Nextcloud. So maybe it is useful for you. If you find a better way, please let me know.

**Comment by alexanderdd** [comment]
related: https://apps.nextcloud.com/apps/integration_davc and https://github.com/nextcloud/integration_davc

**Comment by SebastianKrupinski** [comment]
Closing as completed. 

https://apps.nextcloud.com/apps/integration_davc

**Comment by szaimen** [comment]
🎉🎉🎉🎉🎉
